javatests/arcs/core/util/StateMachineTest.kt

```diff
@@ -0,0 +1,367 @@
+            error(TestExceptionInExit, "Boom!")
+        )
+        assertThat(machine.idle).isTrue()
+    }
+
+    @Test
+    fun testErrorDuringTransition() = runBlockingTest {
+        val machine = TestMachine(TestExceptionInTransition)
+        machine.triggerImmediate(TestEntryAndExitEvent)
+        assertThat(machine.state).isEqualTo(TestExceptionInTransition)
+        assertThat(machine.actionsHandled).containsExactly(
+            exit(TestExceptionInTransition),
+            transition(TestExceptionInTransition, TestState),
+            error(TestExceptionInTransition, "Boom!")
+        )
+        assertThat(machine.idle).isTrue()
+    }
+
+    @Test
+    fun testErrorDuringEntry() = runBlockingTest {
+        val machine = TestMachine()
+        machine.triggerImmediate(TestExceptionInEntryEvent)
+        assertThat(machine.state).isEqualTo(TestExceptionInEntry)
+        assertThat(machine.actionsHandled).containsExactly(
+            exit(Before),
+            transition(Before, TestExceptionInEntry),
+            enter(TestExceptionInEntry),
+            error(TestExceptionInEntry, "Boom!")
+        )
+        assertThat(machine.idle).isTrue()
+    }
+
+    @Test
+    fun testEventQueue() = runBlockingTest {
+        val machine = TestMachine(TestEventQueue)
+        machine.triggerImmediate(TestEventQueueEvent)
+        assertThat(machine.state).isEqualTo(TestEventQueue3)
+        assertThat(machine.actionsHandled).containsExactly(
+            exit(TestEventQueue),
+            transition(TestEventQueue, TestEventQueue2),
+            enter(TestEventQueue2),
+            exit(TestEventQueue2),
+            transition(TestEventQueue2, TestEventQueue3),
+            enter(TestEventQueue3),
+            exit(TestEventQueue3),
+            transition(TestEventQueue3, TestEventQueue3),
+            enter(TestEventQueue3),
+            exit(TestEventQueue3),
+            transition(TestEventQueue3, TestEventQueue3),
+            enter(TestEventQueue3)
+        )
+        assertThat(machine.idle).isTrue()
+    }
+
+    @Test
+    fun testIllegalTransition() = runBlockingTest {
+        val machine = TestMachine(TestEventQueue)
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            machine.triggerImmediate(TestEntryAndExitEvent)
+        }
+    }
+
+
+
+    @Test
+    fun testIllegalMachines() = runBlockingTest {
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupState()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupEntry()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupExit()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupError()
+        }
+        assertSuspendingThrows(IllegalArgumentException::class) {
+            IllegalMachineDupTransition()
+        }
+    }
```

Comment:
It'd be more readable and easier to understand failures if these were split into separate test cases, and the `Illegal*()` things were inlined.